File: 백준/Silver/27514. 1차원 2048/1차원 2048.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
map<ll, ll> mp;

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    int n; cin >> n;
    vector<ll> a(n);
    for (int i = 0; i < n; i++)
    {
        cin >> a[i];
        mp[a[i]]++;
    }

    for (ll num = 1; num <= (1ll << 61); num *= 2)
    {
        mp[num * 2] += mp[num] / 2;
        mp[num] %= 2;
    }

    ll ans = 0;

    for (auto to : mp)
    {
        if (to.second > 0)
            ans = max(ans, to.first);
    }

    cout << ans;
}

File: 백준/Silver/27514. 1차원 2048/1차원 2048.py
n = int(input())
a = list(map(int, input().split()))
cnt = [0 for _ in range(100)]
ans = -1

for i in range(n):
    for j in range(0, 64):
        if a[i] & (1 << j):
            cnt[j] += 1
            ans = max(ans, (1 << j))
            break

num = 0

while num <= 62:
    cnt[num + 1] += cnt[num] // 2
    cnt[num] %= 2
    if cnt[num + 1] > 0:
        ans = (1 << (num + 1))
    num += 1

print(ans)
